sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs

```diff
@@ -0,0 +1,291 @@
+        /// </summary>
+        /// <returns>Bytes read.</returns>
+        /// <remarks>
+        /// This method may have read bytes even if it has reached the confirmed end of stream. You will have to call
+        /// this method again and read zero bytes to get that confirmation.
+        /// </remarks>
+        private static async Task<int> ReadLoopInternal(Stream stream, byte[] buffer, int offset, int minCount, int maxCount, bool async, CancellationToken cancellationToken)
+        {
+            if (minCount > maxCount) // nonsensical; don't allow caller methods to produce this
+            {
+                throw new ArgumentException($"{nameof(minCount)} cannot be greater than {nameof(maxCount)}.");
+            }
+            if (maxCount <= 0) // useless; don't allow caller methods to produce this
+            {
+                throw new ArgumentException("Cannot read a non-positive number of bytes.");
+            }
+
+            int totalRead = 0;
+            do
+            {
+                int read = async
+                    ? await stream.ReadAsync(buffer, offset + totalRead, maxCount - totalRead, cancellationToken).ConfigureAwait(false)
+                    : stream.Read(buffer, offset + totalRead, maxCount - totalRead);
+                if (read == 0) // either we have read maxCount in total or the stream has ended
+                {
+                    break;
+                }
+                totalRead += read;
+            } while (totalRead < minCount); // we always request the number that will bring our total read to maxCount
+                                            // if the stream can only give us so much at the moment and we've at least hit minCount, we can exit
+            return totalRead;
+        }
+
+        public override bool CanRead => true;
+
+        public override bool CanSeek => true;
+
+        public override bool CanWrite => false;
+
+        public override long Length => BufferSet.Sum(tuple => (long)tuple.GarbageOffset);
+
+        public override long Position { get; set; }
+
+        public override void Flush()
+        {
+            // no-op, just like MemoryStream
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            if (Position >= Length)
+            {
+                return 0;
+            }
+
+            int read = 0;
+            while (read < count && Position < Length)
+            {
+                (byte[] currentBuffer, int bufferCount, long offsetOfBuffer) = GetBufferFromPosition();
+
+                int toCopy = (int)Math.Min(Math.Min(
+                    Length - Position,
+                    bufferCount - (Position - offsetOfBuffer)),
+                    count - read);
+                Console.WriteLine($"Offset = {Position - offsetOfBuffer}; Count = {toCopy}; ArrayLength = {currentBuffer.Length}; RecordedBufferCount = {bufferCount};");
```

Comment:
Probably want to remove the Console.WriteLine()

Reply:
oops